### PR TITLE
Add shapes to core/geom

### PR DESCRIPTION
**Pull request by luishfonseca**

### CODE REVIEW

File: core/include/cubos/core/geom/capsule.hpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+            return {radius, 0.0f};
+        }
+
+        /// @return The height of the capsule.
+        float height() const
+        {
+            return lenght + 2.0f * radius;
+        }
+    };
+} // namespace cubos::core::geom
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/capsule.hpp:23]:
```suggestion
} // namespace cubos::core::geom

```

File: core/include/cubos/core/geom/plane.hpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+#include <glm/glm.hpp>
+
+namespace cubos::core::geom
+{
+    struct Plane
+    {
+        const glm::vec3 normal; ///< The normal of the plane.
+    };
+} // namespace cubos::core::geom
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/plane.hpp:11]:
```suggestion
} // namespace cubos::core::geom

```

File: core/include/cubos/core/geom/simplex.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            return {{p1, p2, p3}};
+        }
+
+        static const Simplex tetrahedron(const glm::vec3& p1, const glm::vec3& p2, const glm::vec3& p3,
+                                         const glm::vec3& p4)
+        {
+            return {{p1, p2, p3, p4}};
+        }
+    };
+} // namespace cubos::core::geom
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/simplex.hpp:42]:
```suggestion
} // namespace cubos::core::geom

```

File: core/tests/geom/capsule.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        CHECK_EQ(capsule.height(), 2 * r + l);
+    }
+
+    SUBCASE("sphere")
+    {
+        auto sphere = Capsule::sphere(1.0f);
+        CHECK_EQ(sphere.height(), 2 * r);
+    }
+}
```
Comment (luishfonseca) [inline core/tests/geom/capsule.cpp:23]:
```suggestion
}

```

File: core/tests/geom/simplex.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    SUBCASE("tetrahedron")
+    {
+        auto tetrahedron = Simplex::tetrahedron(p1, p2, p3, p4);
+        CHECK_EQ(tetrahedron.points.size(), 4);
+        CHECK_EQ(tetrahedron.points[0], p1);
+        CHECK_EQ(tetrahedron.points[1], p2);
+        CHECK_EQ(tetrahedron.points[2], p3);
+        CHECK_EQ(tetrahedron.points[3], p4);
+    }
+}
```
Comment (luishfonseca) [inline core/tests/geom/simplex.cpp:53]:
```suggestion
}

```

File: core/tests/geom/utils.hpp [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+#define CHECK_GLM_EQ(a, ...) CHECK_MESSAGE((a == __VA_ARGS__), to_string(a), " != ", to_string(__VA_ARGS__))
+
+template <typename T>
+std::string to_string(const T& t)
+{
+    std::stringstream ss;
+    ss << t;
+    return ss.str();
+}
```
Comment (luishfonseca) [inline core/tests/geom/utils.hpp:13]:
```suggestion
}

```

**Review by luishfonseca — COMMENTED** [review]
missing newlines at eof

File: core/include/cubos/core/geom/capsule.hpp [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#include <vector>
+
+namespace cubos::core::geom
+{
+    struct Capsule
+    {
+        const float radius; ///< The radius of the capsule.
+        const float lenght; ///< The lenght of the capsule.
+
+        static const Capsule sphere(float radius)
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/capsule.hpp:12]: Remove const, doesnt make any difference + add docs

File: core/include/cubos/core/geom/simplex.hpp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            return {{p1, p2}};
+        }
+
+        static const Simplex triangle(const glm::vec3& p1, const glm::vec3& p2, const glm::vec3& p3)
+        {
+            return {{p1, p2, p3}};
+        }
+
+        static const Simplex tetrahedron(const glm::vec3& p1, const glm::vec3& p2, const glm::vec3& p3,
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/simplex.hpp:36]: Unnecessary const

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Also missing file and struct level docs

File: core/include/cubos/core/geom/box.hpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+#include <glm/glm.hpp>
+
+namespace cubos::core::geom
+{
+    /// @brief Box shape.
+    struct Box
+    {
+        const glm::vec3 halfSize; ///< The half size of the box.
+
+        /// Computes two opposite corners of the box.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/box.hpp:15]: Use `@brief`

File: core/include/cubos/core/geom/capsule.hpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
+/// @file
+/// @brief Contains the class for the Capsule shape.
+
+#pragma once
+
+namespace cubos::core::geom
+{
+    /// @brief Capsule shape.
+    /// Can be a capsule or a sphere.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/capsule.hpp:9]:
```suggestion
    /// @brief Capsule shape, which can also represent a sphere.
```

File: core/include/cubos/core/geom/capsule.hpp [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    struct Capsule
+    {
+        const float radius; ///< The radius of the capsule.
+        const float lenght; ///< The lenght of the capsule.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/capsule.hpp:13]:
```suggestion
        const float length; ///< The length of the capsule.
```

File: core/include/cubos/core/geom/simplex.hpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+#pragma once
+
+#include <vector>
+
+#include <glm/glm.hpp>
+
+namespace cubos::core::geom
+{
+    /// @brief A simplex.
+    /// Can be an empty, a point, a line, a triangle or a tetrahedron.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/simplex.hpp:13]:
```suggestion
    /// @brief A simplex shape - may either be empty, a point, a line, a triangle or a tetrahedron.
```

File: core/include/cubos/core/geom/simplex.hpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    struct Simplex
+    {
+        // TODO: We should implement something like boost::static_vector for this. See
+        // https://github.com/GameDevTecnico/cubos/issues/419
+
+        const std::vector<glm::vec3> points; ///< The points of the simplex.
+
+        /// Constructs an empty simplex.
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/simplex.hpp:21]: Use `@brief`

File: core/tests/geom/utils.hpp [hunk 14]
```diff
@@ -0,0 +1,12 @@
+#include <sstream>
+
+#include <glm/glm.hpp>
+
+#define CHECK_VEC3_EQ(a, ...) CHECK_MESSAGE((a == __VA_ARGS__), vec_to_string(a), " != ", vec_to_string(__VA_ARGS__))
+
+std::string vec_to_string(const glm::vec3& v)
```
Comment (RiscadoA) [inline core/tests/geom/utils.hpp:7]:
```suggestion
inline std::string vec_to_string(const glm::vec3& v)
```

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: core/include/cubos/core/geom/box.hpp (hunk 9 above)
Comment (luishfonseca) [inline core/include/cubos/core/geom/box.hpp:15]:
```suggestion
        /// @brief Computes two opposite corners of the box.
```

File: core/include/cubos/core/geom/box.hpp [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        /// @param corners The array to store the two corners in.
+        void corners2(glm::vec3 corners[2]) const
+        {
+            corners[0] = {-halfSize.x, -halfSize.y, -halfSize.z};
+            corners[1] = {halfSize.x, halfSize.y, halfSize.z};
+        }
+
+        /// Computes six opposite corners of the box. Opposite corners are adjacent in the array.
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/box.hpp:23]:
```suggestion
        /// @brief Computes six opposite corners of the box. Opposite corners are adjacent in the array.
```

File: core/include/cubos/core/geom/box.hpp [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            corners[0] = {-halfSize.x, -halfSize.y, -halfSize.z};
+            corners[1] = {halfSize.x, halfSize.y, halfSize.z};
+            corners[2] = {-halfSize.x, -halfSize.y, halfSize.z};
+            corners[3] = {halfSize.x, halfSize.y, -halfSize.z};
+            corners[4] = {-halfSize.x, halfSize.y, -halfSize.z};
+            corners[5] = {halfSize.x, -halfSize.y, halfSize.z};
+        }
+
+        /// Computes the eight corners of the box. Opposite corners are adjacent in the array.
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/box.hpp:35]:
```suggestion
        /// @brief Computes the eight corners of the box. Opposite corners are adjacent in the array.
```

File: core/include/cubos/core/geom/capsule.hpp [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+        /// Constructs a sphere.
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/capsule.hpp:15]:
```suggestion
        /// @brief Constructs a sphere.
```

File: core/include/cubos/core/geom/simplex.hpp (hunk 13 above)
Comment (luishfonseca) [inline core/include/cubos/core/geom/simplex.hpp:21]:
```suggestion
        /// @brief Constructs an empty simplex.
```

File: core/include/cubos/core/geom/simplex.hpp [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        /// @returns The empty simplex.
+        static Simplex empty()
+        {
+            return {{}};
+        }
+
+        /// Constructs a point simplex.
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/simplex.hpp:28]:
```suggestion
        /// @brief Constructs a point simplex.
```

File: core/include/cubos/core/geom/simplex.hpp [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        /// @param p The point.
+        /// @returns The point simplex.
+        static Simplex point(const glm::vec3& p)
+        {
+            return {{p}};
+        }
+
+        /// Constructs a line simplex.
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/simplex.hpp:36]:
```suggestion
        /// @brief Constructs a line simplex.
```

File: core/include/cubos/core/geom/simplex.hpp [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+        /// @param p1 The first point.
+        /// @param p2 The second point.
+        /// @returns The line simplex.
+        static Simplex line(const glm::vec3& p1, const glm::vec3& p2)
+        {
+            return {{p1, p2}};
+        }
+
+        /// Constructs a triangle simplex.
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/simplex.hpp:45]:
```suggestion
        /// @brief Constructs a triangle simplex.
```

File: core/include/cubos/core/geom/simplex.hpp [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+        /// @param p1 The first point.
+        /// @param p2 The second point.
+        /// @param p3 The third point.
+        /// @returns The triangle simplex.
+        static Simplex triangle(const glm::vec3& p1, const glm::vec3& p2, const glm::vec3& p3)
+        {
+            return {{p1, p2, p3}};
+        }
+
+        /// Constructs a tetrahedron simplex.
```
Comment (luishfonseca) [inline core/include/cubos/core/geom/simplex.hpp:55]:
```suggestion
        /// @brief Constructs a tetrahedron simplex.
```

**Review by RiscadoA — APPROVED** [review]

**Review by RiscadoA — APPROVED** [review]

**Pull request merged**